src/main/scala/is/hail/utils/IntervalTree.scala

```diff
@@ -10,117 +10,128 @@ import scala.collection.mutable
+  * precondition, that there exists a Hail type 't: Type' such that 'pord' was
+  * constructed by 't.ordering', and either
+  * - 't: TBaseStruct' and 't.relaxedTypeCheck(x)', or
+  * - 't.typeCheck(x)',
+  * where 'x' is each of 'left', 'right', 'p', 'other.left', and 'other.right'
+  * as appropriate. In the case 't: TBaseStruct', 't' could be replaced by any
+  * 't2' such that 't.isPrefixOf(t2)' without changing the behavior.
+  */
 case class Interval(start: Any, end: Any, includesStart: Boolean, includesEnd: Boolean) extends Serializable {
+  val (left, right) = Interval.toIntervalEndpoints(start, end, includesStart, includesEnd)
 
-  def contains(pord: ExtendedOrdering, position: Any): Boolean = {
-    val compareStart = pord.compare(position, start)
-    if (compareStart > 0 || (compareStart == 0 && includesStart)) {
-      val compareEnd = pord.compare(position, end)
-      compareEnd < 0 || (compareEnd == 0 && includesEnd)
-    } else
-      false
-  }
+  private def ext(pord: ExtendedOrdering): ExtendedOrdering = pord.intervalEndpointOrdering
+
+  def contains(pord: ExtendedOrdering, p: Any): Boolean =
+    ext(pord).lt(left, p) && ext(pord).gt(right, p)
 
   def includes(pord: ExtendedOrdering, other: Interval): Boolean =
-    other.definitelyEmpty(pord) || ({
-      val cstart = pord.compare(this.start, other.start)
-      cstart < 0 || (cstart == 0 && (!other.includesStart || this.includesStart))
-    } && {
-      val cend = pord.compare(this.end, other.end)
-      cend > 0 || (cend == 0 && (!other.includesEnd || this.includesEnd))
-    })
-
-  def mayOverlap(pord: ExtendedOrdering, other: Interval): Boolean = {
-    !definitelyDisjoint(pord, other)
-  }
+    ext(pord).lteq(this.left, other.left) && ext(pord).gteq(this.right, other.right)
+
+  def mayOverlap(pord: ExtendedOrdering, other: Interval): Boolean =
+    ext(pord).lt(this.left, other.right) && ext(pord).gt(this.right, other.left)
 
   def isAbovePosition(pord: ExtendedOrdering, p: Any): Boolean =
-    definitelyEmpty(pord) || {
-      val c = pord.compare(p, start)
-      c < 0 || (c == 0 && !includesStart)
-    }
+    ext(pord).gt(left, p)
 
   def isBelowPosition(pord: ExtendedOrdering, p: Any): Boolean =
-    definitelyEmpty(pord) || {
-      val c = pord.compare(p, end)
-      c > 0 || (c == 0 && !includesEnd)
-    }
+    ext(pord).lt(right, p)
 
   def definitelyDisjoint(pord: ExtendedOrdering, other: Interval): Boolean =
-    definitelyEmpty(pord) || other.definitelyEmpty(pord) ||
-      isBelow(pord, other) || isAbove(pord, other)
-
-  private var _emptyDefined: Boolean = false
-  private var _empty: Boolean = false
+    !mayOverlap(pord, other)
 
-  def definitelyEmpty(pord: ExtendedOrdering): Boolean = {
-    if (!_emptyDefined) {
-      _empty = if (includesStart && includesEnd) pord.gt(start, end) else pord.gteq(start, end)
-      _emptyDefined = true
-    }
-    _empty
-  }
+  def definitelyEmpty(pord: ExtendedOrdering): Boolean = false
```

Comment:
is this still called anywhere? can we remove it?

Reply:
Yes!